Data Wrangling

In [178]:
import pandas as pd
import matplotlib.pylab as plt

df = pd.read_csv('wandaloo_cars.csv', delimiter=',')
df.head()

Brand           Model        Version       Price  Modèle     Main  \
0  MERCEDES   Classe C 2014          220 d  228.000 DH  2014.0  Seconde   
1  MERCEDES   Classe C 2003        220 CDI   79.000 DH  2003.0  Seconde   
2  MERCEDES   Classe B 2013    180 d Urban  150.000 DH  2013.0  Seconde   
3  MERCEDES   Classe E 2006  Autre version  135.000 DH  2006.0  Seconde   
4  MERCEDES  CLA Coupé 2014        220 CDI  240.000 DH  2014.0  Seconde   

  Kilométrage Carburant  Transmision Puissance fiscale  ... Airbags  ABS  ESP  \
0  144.500 km    Diesel  Automatique              8 cv  ...     oui  oui  oui   
1  350.000 km    Diesel     Manuelle              8 cv  ...     oui  oui  oui   
2  144.000 km    Diesel     Manuelle              8 cv  ...     oui  oui  oui   
3  248.000 km    Diesel  Automatique              9 cv  ...     oui  oui  oui   
4  103.000 km    Diesel  Automatique              9 cv  ...     oui  oui  oui   

  Antipatinage Architecture Cylindrée Conso. ville Conso. route Vitesse maxi.  \
0          oui          NaN       NaN          NaN          NaN           NaN   
1          oui          NaN       NaN          NaN          NaN           NaN   
2          oui          NaN       NaN          NaN          NaN           NaN   
3          oui          NaN       NaN          NaN          NaN           NaN   
4          oui          NaN       NaN          NaN          NaN           NaN   

  Volume du réservoir  
0                 NaN  
1                 NaN  
2                 NaN  
3                 NaN  
4                 NaN  

[5 rows x 43 columns]

Steps for working with missing data:

Identify missing data

1.Deal with missing data
2.Correct data format
3.Identify missing values

In the car dataset, missing data comes with NaN (Not a Number), which is Python's default missing value marker for reasons of computational speed and convenience.

Evaluating for Missing Data
The missing values are converted to Python's default. We use Python's built-in functions to identify these missing values. There are two methods to detect missing data:

1/ .isnull()
2/ .notnull()
The output is a boolean value indicating whether the value that is passed into the argument is in fact missing data

In [47]:
missing_data = df.isnull()
missing_data.head(5)

Brand  Model  Version  Price  Modèle   Main  Kilométrage  Carburant  \
0  False  False    False  False   False  False        False      False   
1  False  False    False  False   False  False        False      False   
2  False  False    False  False   False  False        False      False   
3  False  False    False  False   False  False        False      False   
4  False  False    False  False   False  False        False      False   

   Transmision  Puissance fiscale  ...  Airbags    ABS    ESP  Antipatinage  \
0        False              False  ...    False  False  False         False   
1        False              False  ...    False  False  False         False   
2        False              False  ...    False  False  False         False   
3        False              False  ...    False  False  False         False   
4        False              False  ...    False  False  False         False   

   Architecture  Cylindrée  Conso. ville  Conso. route  Vitesse maxi.  \
0          True       True          True          True           True   
1          True       True          True          True           True   
2          True       True          True          True           True   
3          True       True          True          True           True   
4          True       True          True          True           True   

   Volume du réservoir  
0                 True  
1                 True  
2                 True  
3                 True  
4                 True  

[5 rows x 43 columns]

"True" stands for missing value, while "False" stands for not missing value.

Count missing values in each column
Using a for loop in Python, we can quickly figure out the number of missing values in each column."True" represents a missing value, "False" means the value is present in the dataset. In the body of the for loop the method ".value_counts()" counts the number of "True" values.

In [48]:
for column in missing_data.columns.values.tolist():
    print(column)
    print(missing_data[column].value_counts())
    print("") 

Brand
False    3063
True       66
Name: Brand, dtype: int64

Model
False    3063
True       66
Name: Model, dtype: int64

Version
False    2180
True      949
Name: Version, dtype: int64

Price
False    3063
True       66
Name: Price, dtype: int64

Modèle
False    2180
True      949
Name: Modèle, dtype: int64

Main
False    2180
True      949
Name: Main, dtype: int64

Kilométrage
False    2180
True      949
Name: Kilométrage, dtype: int64

Carburant
False    2166
True      963
Name: Carburant, dtype: int64

Transmision
False    2180
True      949
Name: Transmision, dtype: int64

Puissance fiscale
False    3048
True       81
Name: Puissance fiscale, dtype: int64

Couleur extérieure
False    2180
True      949
Name: Couleur extérieure, dtype: int64

Etat du véhicule
False    2180
True      949
Name: Etat du véhicule, dtype: int64

Climatisation
False    3048
True       81
Name: Climatisation, dtype: int64

Vitres électriques
False    3048
True       81
Name: Vitres électriques, dtype: int

Each column has 3128 rows of data and 43 columns containing missing data:

"Modèle": 66 missing data
"Price": 66 missing data
"Puissance fiscale": 81 missing data
"Cylindrée" : 2284 missing data
"Conso. route": 2284 missing data
"Vitesse maxi.": 2261 missing data
"Volume du réservoir": 2284 missing data
.
.
.

Deal with missing data
1.drop data
    a. drop the whole row
    b. drop the whole column
2.replace data
    a. replace it by mean
    b. replace it by frequency
    c. replace it based on other functions

Whole columns should be dropped only if most entries in the column are empty. In our dataset, some columns are empty enough to drop entirely (Architecture 2284 missing data
Cylindrée 2284 missing data
conso. ville 2284 missing data
conso. route 2284 missing data
Vitesse maxi. 2261 missing data
Volume du réservoire 2284 missing data).
We have some freedom in choosing which method to replace data; however, some methods may seem more reasonable than others. We will apply each method to many different columns:

Replace by mean:

    "Modèle": 949 missing data, replace them with mean
    

Replace by frequency:

    "Main": 949 missing data, replace them with "Seconde".
        Reason: 60% sedans is seconde main. Since seconde main is most frequent, it is most likely to occur
    "Carburant": 963 missing data, replace them with "Diesel"
        Reason: 90% sedans is diesel carburant
    "Transmision": 949 missing data, replace them with "Manuelle"
    "Couleur extérieure": 949 missing data, replace them with "Noir"
    "Climatisation" 81 missing data, replace them with ""
    Vitre électriques: 81 missing data, replace them with ""
    "Sièges électriques": 101 missing data, replace them with "non"
    "Ordinateur de board":
    "Allumage auto. des feux":
    "Start & Stop"; replace missing data with "oui"
Drop the whole row:

"price": 66 missing data, simply delete the whole row
        Reason: price is what we want to predict. Any data entry without price data cannot be used for prediction; therefore any row now without price data is not useful to us
"Brand", "Model", "Version", "Kilométrage", "Puissance fiscale", "Etat du véhicule"
Drop the whole columns:
(Architecture 2284 missing data
Cylindrée 2284 missing data
conso. ville 2284 missing data
conso. route 2284 missing data
Vitesse maxi. 2261 missing data
Volume du réservoire 2284 missing data).
    Reason: most entries in the columns are empty

Calculate the average of the columns : "Modèle", 'Puissance fiscale'

In [49]:
avg_modèle = df["Modèle"].astype("float").mean(axis=0)
print("Average of modèle:", avg_modèle)

Average of modèle: 2012.0022935779816


In [50]:
import numpy as np
df["Modèle"].replace(np.nan,avg_modèle,inplace=True)

To see which values are present in a particular column,use the ".value_counts()" method:

In [52]:
df['Main'].value_counts()

Seconde     1126
Première    1054
Name: Main, dtype: int64

We can see that seconde main is the most common type. We can also use the ".idxmax()" method to calculate for us the most common type automatically:



In [55]:
df['Main'].value_counts().idxmax()


'Diesel'

In [54]:
#replace the missing 'Main' values by the most frequent 
df['Main'].replace(np.nan,"Seconde",inplace=True)

In [57]:
df['Carburant'].value_counts()

Diesel        1902
Essence        256
Hybride          7
Electrique       1
Name: Carburant, dtype: int64

In [56]:
df['Carburant'].value_counts().idxmax()

'Diesel'

In [ ]:
#replace the missing 'Crburant' values by the most frequent 
df['Carburant'].replace(np.nan,"Diesel",inplace=True)

In [58]:
df['Transmision'].value_counts()

Manuelle       1146
Automatique    1034
Name: Transmision, dtype: int64

In [59]:
#replace the missing 'Transmision' values by the most frequent 
df['Transmision'].replace(np.nan,"Manuelle",inplace=True)

In [68]:
df['Couleur extérieure'].value_counts()

Noir             617
Blanc            280
Gris foncé       224
Gris             222
-                219
Gris clair       155
Bleu foncé        87
Autre couleur     67
Marron            50
Bleu              49
Rouge             47
Beige             31
Blanc cassé       24
Bleu ciel         23
Bordeau           21
Marron clair      17
Vert foncé        13
Jaune d'or        10
Violet             9
Vert               9
Orange             2
Jaune              2
Rose               2
Name: Couleur extérieure, dtype: int64

In [69]:
#replace the missing 'Couleur extérieure' values by the most frequent 
df['Couleur extérieure'].replace(np.nan,"Noir",inplace=True)

In [71]:
df['Climatisation'].value_counts()


auto. Bi-zone        946
manuelle             926
auto.                513
non                  409
auto. Tri-zone       141
auto. Multi-zone      59
auto. Quadri-zone     53
تلقاءي.                1
Name: Climatisation, dtype: int64

In [72]:
#replace the missing 'Climatisation' values by the most frequent 
df['Climatisation'].replace(np.nan,"auto. Bi-zone",inplace=True)

In [73]:
df['Vitres électriques'].value_counts()


Electrique AV/AR    1935
Electrique AV        657
-                    445
non                    8
Electrique AR          3
Name: Vitres électriques, dtype: int64

In [74]:
#replace the missing 'Vitres électriques' values by the most frequent 
df['Vitres électriques'].replace(np.nan,"Electrique AV/AR",inplace=True)

In [75]:
df['Sièges électriques'].value_counts()


non                       2006
Conducteur et Passager     594
Conducteur                 340
Tous les passagers          76
Passager                    12
Name: Sièges électriques, dtype: int64

In [76]:
#replace the missing 'Vitres électriques' values by the most frequent 
df['Sièges électriques'].replace(np.nan,"non",inplace=True)

In [77]:
df['Ordinateur de bord'].value_counts()



oui    2192
non     856
Name: Ordinateur de bord, dtype: int64

In [78]:
#replace the missing 'Ordinateur de bord' values by the most frequent 
df['Ordinateur de bord'].replace(np.nan,"oui",inplace=True)

In [79]:
df['Start & Stop'].value_counts()



non    1775
oui    1273
Name: Start & Stop, dtype: int64

In [ ]:
#replace the missing 'Start & Stop' values by the most frequent 
df['Start & Stop'].replace(np.nan,"oui",inplace=True)

In [80]:
df['Régulateur de vitesse'].value_counts()


oui    2186
non     862
Name: Régulateur de vitesse, dtype: int64

In [81]:
#replace the missing 'Régulateur de vitesse' values by the most frequent 
df['Régulateur de vitesse'].replace(np.nan,"oui",inplace=True)

In [82]:
df['Allumage auto. des feux'].value_counts()


oui    1742
non    1306
Name: Allumage auto. des feux, dtype: int64

In [83]:
#replace the missing 'Allumage auto. des feux' values by the most frequent 
df['Allumage auto. des feux'].replace(np.nan,"oui",inplace=True)

In [84]:
df['Détecteur de pluie'].value_counts()


oui    1590
non    1458
Name: Détecteur de pluie, dtype: int64

In [85]:
#replace the missing 'Détecteur de pluie' values by the most frequent 
df['Détecteur de pluie'].replace(np.nan,"oui",inplace=True)

In [86]:
df['Commandes au volant'].value_counts()

oui    2132
non     916
Name: Commandes au volant, dtype: int64

In [ ]:
#replace the missing 'Commandes au volant' values by the most frequent 
df['Commandes au volant'].replace(np.nan,"oui",inplace=True)

In [87]:

df['Ecran tactile'].value_counts()

oui    1662
non    1386
Name: Ecran tactile, dtype: int64

In [88]:
#replace the missing 'Ecran tactile' values by the most frequent 
df['Ecran tactile'].replace(np.nan,"oui",inplace=True)

In [89]:
df['Rétroviseurs électriques'].value_counts()

oui    2139
non     909
Name: Rétroviseurs électriques, dtype: int64

In [ ]:
#replace the missing 'Rétroviseurs électriques' values by the most frequent 
df['Rétroviseurs électriques'].replace(np.nan,"oui",inplace=True)

In [90]:
df['Démarrage mains libres'].value_counts()

non    1666
oui    1382
Name: Démarrage mains libres, dtype: int64

In [91]:
#replace the missing 'Démarrage mains libres' values by the most frequent 
df['Démarrage mains libres'].replace(np.nan,"non",inplace=True)

In [92]:
df['Banquette arrière rabattable 1/3-2/3'].value_counts()

oui    1655
non    1393
Name: Banquette arrière rabattable 1/3-2/3, dtype: int64

In [93]:
#replace the missing 'Banquette arrière rabattable 1/3-2/3' values by the most frequent 
df['Banquette arrière rabattable 1/3-2/3'].replace(np.nan,"oui",inplace=True)

In [94]:
df['Caméra de recul'].value_counts()

non    1706
oui    1342
Name: Caméra de recul, dtype: int64

In [95]:
#replace the missing 'Caméra de recul' values by the most frequent 
df['Caméra de recul'].replace(np.nan,"non",inplace=True)

In [96]:
df['Bluetooth'].value_counts()

oui    2116
non     932
Name: Bluetooth, dtype: int64

In [97]:
#replace the missing 'Bluetooth' values by the most frequent 
df['Bluetooth'].replace(np.nan,"oui",inplace=True)

In [98]:
df['Jantes aluminium'].value_counts()

oui          1429
non           837
18 pouces     198
17 pouces     187
19 pouces     155
16 pouces     107
20 pouces      76
21 pouces      24
15 pouces      22
14 pouces       8
22 pouces       5
Name: Jantes aluminium, dtype: int64

In [99]:
#replace the missing 'Jantes aluminium' values by the most frequent 
df['Jantes aluminium'].replace(np.nan,"oui",inplace=True)

In [100]:
df['Volant cuir'].value_counts()

non    1830
oui    1218
Name: Volant cuir, dtype: int64

In [101]:
#replace the missing 'Volant cuir' values by the most frequent 
df['Volant cuir'].replace(np.nan,"non",inplace=True)

In [102]:
df['Feux de jour'].value_counts()

oui    1870
non    1178
Name: Feux de jour, dtype: int64

In [ ]:
#replace the missing 'Feux de jour' values by the most frequent 
df['Feux de jour'].replace(np.nan,"oui",inplace=True)

In [103]:
df['Barres de toit'].value_counts()

non    2316
oui     732
Name: Barres de toit, dtype: int64

In [ ]:
#replace the missing 'Barres de toit' values by the most frequent 
df['Barres de toit'].replace(np.nan,"non",inplace=True)

In [104]:
df['Toit'].value_counts()

Rigide                 2394
Panoramique ouvrant     362
Panoramique             198
Ouvrant                  94
Name: Toit, dtype: int64

In [105]:
#replace the missing 'Toit' values by the most frequent 
df['Toit'].replace(np.nan,"Rigide",inplace=True)

In [106]:
df['Airbags'].value_counts()

oui     1922
6.0      290
non      258
8.0      183
7.0      171
2.0      117
4.0       51
9.0       24
10.0      23
0.0        7
1.0        2
Name: Airbags, dtype: int64

In [ ]:
#replace the missing 'Airbags' values by the most frequent 
df['Airbags'].replace(np.nan,"oui",inplace=True)

In [107]:
df['ABS'].value_counts()

oui    2704
non     344
Name: ABS, dtype: int64

In [108]:
#replace the missing 'ABS' values by the most frequent 
df['ABS'].replace(np.nan,"oui",inplace=True)

In [109]:
df['ESP'].value_counts()

oui    2235
non     813
Name: ESP, dtype: int64

In [110]:
#replace the missing 'ESP' values by the most frequent 
df['ESP'].replace(np.nan,"oui",inplace=True)

In [111]:
df['Antipatinage'].value_counts()

oui    2094
non     954
Name: Antipatinage, dtype: int64

In [112]:
#replace the missing 'Antipatinage' values by the most frequent 
df['Antipatinage'].replace(np.nan,"oui",inplace=True)

In [51]:
# Delete some columns by name ['Cylindrée', 'Conso. ville', 'Conso. route', 'Vitesse maxi.', 'Volume du réservoir']
df = df.drop(['Architecture', 'Cylindrée', 'Conso. ville', 'Conso. route', 'Vitesse maxi.', 'Volume du réservoir'], axis=1)
print(df)

         Brand           Model        Version                           Price  \
0     MERCEDES   Classe C 2014          220 d                      228.000 DH   
1     MERCEDES   Classe C 2003        220 CDI                       79.000 DH   
2     MERCEDES   Classe B 2013    180 d Urban                      150.000 DH   
3     MERCEDES   Classe E 2006  Autre version                      135.000 DH   
4     MERCEDES  CLA Coupé 2014        220 CDI                      240.000 DH   
...        ...             ...            ...                             ...   
3124     VOLVO            XC90            NaN    630.000 DH *   * Prix public   
3125     VOLVO            XC90            NaN    730.000 DH *   * Prix public   
3126     VOLVO            XC90            NaN    850.000 DH *   * Prix public   
3127     VOLVO            XC90            NaN  1.000.000 DH *   * Prix public   
3128     VOLVO            XC90            NaN  1.100.000 DH *   * Prix public   

           Modèle     Main 

Finally, let's drop all rows that do not have price, brand, model... data:

In [163]:
# simply drop whole row with NaN in "price", "Brand", "Model", "Version" columns
df.dropna(subset=["Price"],axis=0,inplace=True)
df.dropna(subset=["Brand"],axis=0,inplace=True)
df.dropna(subset=["Model"],axis=0,inplace=True)
df.dropna(subset=["Version"],axis=0,inplace=True)
df.dropna(subset=["Kilométrage"],axis=0,inplace=True)
df.dropna(subset=["Puissance fiscale"],axis=0,inplace=True)
df.dropna(subset=["Etat du véhicule"],axis=0,inplace=True)
df.dropna(subset=["Ouverture auto. du coffre"],axis=0,inplace=True)


# reset index, because we droped two rows
df.reset_index(drop=True,inplace=True)

In [114]:
df.head()

Brand           Model        Version       Price  Modèle     Main  \
0  MERCEDES   Classe C 2014          220 d  228.000 DH  2014.0  Seconde   
1  MERCEDES   Classe C 2003        220 CDI   79.000 DH  2003.0  Seconde   
2  MERCEDES   Classe B 2013    180 d Urban  150.000 DH  2013.0  Seconde   
3  MERCEDES   Classe E 2006  Autre version  135.000 DH  2006.0  Seconde   
4  MERCEDES  CLA Coupé 2014        220 CDI  240.000 DH  2014.0  Seconde   

  Kilométrage Carburant  Transmision Puissance fiscale  ... Bluetooth  \
0  144.500 km    Diesel  Automatique                    ...       oui   
1  350.000 km    Diesel     Manuelle                    ...       oui   
2  144.000 km    Diesel     Manuelle                    ...       oui   
3  248.000 km    Diesel  Automatique                    ...       oui   
4  103.000 km    Diesel  Automatique                    ...       oui   

  Jantes aluminium Volant cuir Feux de jour Barres de toit         Toit  \
0              oui         oui          oui            non       Rigide   
1              oui         non          oui            non       Rigide   
2              oui         oui          oui            non       Rigide   
3              oui         non          oui            oui       Rigide   
4              oui         non          oui            non  Panoramique   

  Airbags  ABS  ESP Antipatinage  
0     oui  oui  oui          oui  
1     oui  oui  oui          oui  
2     oui  oui  oui          oui  
3     oui  oui  oui          oui  
4     oui  oui  oui          oui  

[5 rows x 37 columns]

In [179]:
df['Price'] = df['Price'].str.replace(' VENDUE','')
df['Price'] = df['Price'].str.replace(' DH','')
df['Price'] = df['Price'].str.replace(' * promo','')
df['Price'] = df['Price'].str.replace('.','')
df['Price'] = df['Price'].str.replace(' *','')
df['Price'] = df['Price'].str.replace('*','')
df['Price'] = df['Price'].str.replace('Prixpublic','')


df['Price'] = df['Price'].astype(float)

C:\Users\GeekStore\AppData\Local\Temp\ipykernel_8388\3965877181.py:3: FutureWarning: The default value of regex will change from True to False in a future version.
  df['Price'] = df['Price'].str.replace(' * promo','')
C:\Users\GeekStore\AppData\Local\Temp\ipykernel_8388\3965877181.py:4: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  df['Price'] = df['Price'].str.replace('.','')
C:\Users\GeekStore\AppData\Local\Temp\ipykernel_8388\3965877181.py:5: FutureWarning: The default value of regex will change from True to False in a future version.
  df['Price'] = df['Price'].str.replace(' *','')
C:\Users\GeekStore\AppData\Local\Temp\ipykernel_8388\3965877181.py:6: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal s

In [ ]:
df.head()

Now, the dataset with no missing values is obtained.

Lets list the data types for each column

In [180]:
df.dtypes

Brand                                    object
Model                                    object
Version                                  object
Price                                   float64
Modèle                                  float64
Main                                     object
Kilométrage                              object
Carburant                                object
Transmision                              object
Puissance fiscale                        object
Couleur extérieure                       object
Etat du véhicule                         object
Climatisation                            object
Vitres électriques                       object
Sièges électriques                       object
Ordinateur de bord                       object
Start & Stop                             object
Régulateur de vitesse                    object
Allumage auto. des feux                  object
Détecteur de pluie                       object
Commandes au volant                     